nixos/modules/programs/wayland/uwsm.nix

```diff
@@ -0,0 +1,117 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.programs.uwsm;
+
+  # Helper function to create desktop entry files for UWSM-managed compositors
+  mk_uwsm_desktop_entry =
+    opts:
+    (pkgs.writeTextFile {
+      name = "${opts.name}_uwsm";
+      text = ''
+        [Desktop Entry]
+        Name=${opts.compositorPrettyName} (UWSM)
+        Comment=${opts.compositorComment}
+        Exec=${lib.getExe cfg.package} start -S -F "${opts.compositorBinPath}"
+        Type=Application
+      '';
+      destination = "/share/wayland-sessions/${opts.name}-uwsm.desktop";
+      derivationArgs = {
+        passthru.providedSessions = [ "${opts.name}-uwsm" ];
+      };
+    });
+in
+{
+  options.programs.uwsm = {
+    enable = lib.mkEnableOption ''
+      uwsm, which wraps standalone Wayland compositors with a set
+      of Systemd units on the fly. This essentially
+      binds the wayland compositor into `graphical-session-pre.target`,
+      `graphical-session.target`, `xdg-desktop-autostart.target`.
+
+      This is useful for Wayland compositors like Hyprland, Sway, Wayfire,
+      etc. that do not start these targets and services on their own.
+
+      If you are having trouble starting a service that depends on
+      `graphical-session.target`, while using a WM, enabling and
+      configuring this might help
+    '';
+    package = lib.mkPackageOption pkgs "uwsm" { };
+
+    waylandCompositors = lib.mkOption {
+      description = ''
+        Configuration for UWSM-managed Wayland Compositors.
+        This creates a desktop entry file which will be used by
+        Display Managers to allow starting the UWSM managed session.
+      '';
+      type = lib.types.attrsOf (
+        lib.types.submodule (
+          { ... }:
+          {
+            options = {
+              # name,
+              # Name, Comment (for desktop entry)
+              # Compositor-Package -> The main binary could be extracted from getExe
+              # which would then be used in `Exec` of desktop entry
+              compositorPrettyName = lib.mkOption {
+                type = lib.types.str;
+                description = "The full name of the desktop entry file.";
+                example = "ExampleWaylandCompositor";
+              };
+              compositorComment = lib.mkOption {
+                type = lib.types.str;
+                description = "The comment field of the desktop entry file.";
+                example = "An intelligent dynamic tiling Wayland compositor managed by UWSM.";
+              };
+              compositorBinPath = lib.mkOption {
+                type = lib.types.path;
+                description = ''
+                  The wayland-compositor binary path that will be called by UWSM.
```

Comment:
Here, I would maybe provide a pointer as to why we are using `/run/current-system/sw/bin` instead of `lib.getExe pkgs.Hyprland` for example.
But I don't know if this would be even more confusing or not.
I am just worried that there would be a weird corner case where somebody uses `lib.getExe pkgs.Wayfire` or similar and have weird issues because of missing plugins.

Reply:
Done.